magma/backend/mlir/when.py

```diff
@@ -0,0 +1,241 @@
+import contextlib
+import itertools
+
+from magma.backend.mlir.builtin import builtin
+from magma.backend.mlir.hw import hw
+from magma.backend.mlir.magma_common import (
+    visit_value_or_value_wrapper_by_direction as
+    visit_magma_value_or_value_wrapper_by_direction,
+)
+from magma.backend.mlir.mlir import get_block_stack, MlirValue, push_block
+from magma.backend.mlir.sv import sv
+from magma.common import sort_by_value
+from magma.primitives.when import iswhen
+from magma.ref import ArrayRef, TupleRef
+from magma.value_utils import make_selector
+
+
+def _assert_false(*args, **kwargs):
+    assert False
+
+
+class WhenCompiler:
+    def __init__(self, module_visitor, module):
+        self._module_visitor = module_visitor
+
+        inst = module.module
+        defn = type(inst)
+        assert iswhen(defn)
+        builder = defn._builder_
+
+        input_to_index = {}
+        output_to_index = {}
+
+        def _visit(value, counter, value_to_index):
```

Comment:
I feel like a value of making this a class is to use member variables to avoid a closure? otherwise I think this should just be a free function -- no point in making it a class

Reply:
nvm i see the stacked PRs